File: Java/Leetcode/MathAndGeo/SetMatrixZeroes.java
package com.brianprojects.Leetcode.MathAndGeo;

import java.util.ArrayDeque;
import java.util.Queue;

public class SetMatrixZeroes {
    //Time Complexity: O(RC) || Space Complexity:O(1)
    public void setZeroes(int[][] matrix){
        if(matrix == null){return;}
        Queue<int[]> queue = new ArrayDeque<>();
        int r = matrix.length;
        int c = matrix[0].length;
        for(int i =0; i<r;i++){
            for(int j = 0; j<c; j++){
                if(matrix[i][j] == 0){
                    queue.add(new int[]{i,j});
                }
            }
        }
        while(!queue.isEmpty()){
            int[] arr = queue.poll();
            fillCzeroes(matrix,arr[0]);
            fillRzeroes(matrix,arr[1]);
        }
    }

    private void fillCzeroes(int[][] matrix, int index) {
        int c = matrix[index].length;
        for(int i =0; i<c;i++){
            matrix[index][i] = 0;
        }
    }

    private void fillRzeroes(int[][] matrix,int index) {

        int r = matrix.length;
        for(int i = 0; i<r;i++){
            matrix[i][index] = 0;
        }

    }


}

File: Python/Leetcode_Python/MathandGeo/SetMatrixZeroes.py
# Time Complexity: O(mn) || Space Complexity: O(mn)
from typing import List


def setZeroes(matrix: List[List[int]]) -> None:
    if not matrix:
        return
    r = len(matrix)
    c = len(matrix[0])

    cords = []

    for i in range(r):
        for j in range(c):
            if matrix[i][j] == 0:
                cords.append([i, j])

    for cord in cords:
        up(matrix, cord[0] - 1, cord[1], r, c)
        down(matrix, cord[0] + 1, cord[1], r, c)
        left(matrix, cord[0], cord[1] - 1, r, c)
        right(matrix, cord[0], cord[1] + 1, r, c)


def up(matrix: List[List[int]], x: int, y: int, r: int, c: int) -> None:
    if x < 0 or y < 0 or x >= r or y >= c:
        return

    matrix[x][y] = 0
    up(matrix, x - 1, y, r, c)


def down(matrix: List[List[int]], x: int, y: int, r: int, c: int) -> None:
    if x < 0 or y < 0 or x >= r or y >= c:
        return

    matrix[x][y] = 0
    down(matrix, x + 1, y, r, c)


def left(matrix: List[List[int]], x: int, y: int, r: int, c: int) -> None:
    if x < 0 or y < 0 or x >= r or y >= c:
        return

    matrix[x][y] = 0
    left(matrix, x, y - 1, r, c)


def right(matrix: List[List[int]], x: int, y: int, r: int, c: int) -> None:
    if x < 0 or y < 0 or x >= r or y >= c:
        return

    matrix[x][y] = 0
    right(matrix, x, y + 1, r, c)
